### PR TITLE
[Refactor] reduce memory usage of array_agg/group_concat

### PR DESCRIPTION
**Pull request by fzhedu**
Why I'm doing:

What I'm doing:

reduce one time of memory copy in finalize in aggregation, totally 1/3 of previous memory costs, besides reducing time up to 1/2. It aims to optimizing low-cardinality agg.

1. early release memory
2. avoid materialize sorted result

```
select sum(length(o)) from (select group_concat(o_orderkey order by o_custkey) as o from orders group by o_shippriority)A
```
  | opt | baseline
-- | -- | --
Cost time | 22s943ms | 45s935ms
QueryPeakMemoryUsage | 8.745GB  | 10.360GB
InstancePeakMemoryUsage | 7.430 | 9.044
AGGREGATE_BLOCKING_SOURCE (plan_node_id=3) | 2.031 | 3.395
AGGREGATE_BLOCKING_SINK (plan_node_id=3) | 2.633 | 2.883

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/exprs/agg/array_agg.h [hunk 1]
```diff
@@ -320,7 +320,13 @@ class ArrayAggAggregateFunctionV2
             }
             auto& offsets = array_col->offsets_column()->get_data();
             offsets.push_back(offsets.back() + elem_size);
+#ifndef BE_TEST
```
Comment (trueeyu) [inline be/src/exprs/agg/array_agg.h:323]: Why not reset for TEST?

Comment (fzhedu) [reply be/src/exprs/agg/array_agg.h]: change ut.

File: be/src/exprs/agg/array_agg.h [hunk 2]
```diff
@@ -213,28 +213,27 @@ struct ArrayAggAggregateStateV2 {
...
+        if (!data_columns.empty()) {
+            for (auto& col : data_columns) {
                 col.reset();
             }
-            data_columns->clear();
-            data_columns.reset(nullptr);
+            data_columns.clear();
         }
     }
     // using pointer rather than vector to avoid variadic size
```
Comment (trueeyu) [inline be/src/exprs/agg/array_agg.h:234]: What's the comment mean?

File: be/src/exprs/agg/array_agg.h [hunk 3]
```diff
@@ -213,28 +213,27 @@ struct ArrayAggAggregateStateV2 {
...
+            data_columns[i].reset();
         }
-        data_columns->resize(1);
+        data_columns.resize(1);
     }
 
     ~ArrayAggAggregateStateV2() {
-        if (data_columns != nullptr) {
-            for (auto& col : *data_columns) {
+        if (!data_columns.empty()) {
```
Comment (trueeyu) [inline be/src/exprs/agg/array_agg.h:227]: Why not only run data_columns.clear()?

File: be/src/exprs/agg/array_agg.h [hunk 4]
```diff
@@ -271,7 +269,7 @@ class ArrayAggAggregateFunctionV2
                 this->data(state).update_nulls(i, 1);
                 continue;
             }
-            auto* data_col = columns[i];
+            auto*& data_col = columns[i];
```
Comment (trueeyu) [inline be/src/exprs/agg/array_agg.h:272]: auto* ?

File: be/src/exprs/agg/array_agg.h [hunk 5]
```diff
@@ -213,28 +213,24 @@ struct ArrayAggAggregateStateV2 {
...
     }
 
     ~ArrayAggAggregateStateV2() {
-        if (data_columns != nullptr) {
-            for (auto& col : *data_columns) {
-                col.reset();
-            }
-            data_columns->clear();
-            data_columns.reset(nullptr);
+        if (!data_columns.empty()) {
```
Comment (trueeyu) [inline be/src/exprs/agg/array_agg.h:227]: no need to check empty?

File: be/src/exprs/agg/array_agg.h [hunk 6]
```diff
@@ -213,28 +213,24 @@ struct ArrayAggAggregateStateV2 {
...
-        for (auto i = 1; i < data_columns->size(); ++i) {
-            data_columns->at(i).reset();
+        for (auto i = 1; i < data_columns.size(); ++i) {
+            data_columns[i].reset();
         }
-        data_columns->resize(1);
+        data_columns.resize(1);
     }
 
     ~ArrayAggAggregateStateV2() {
```
Comment (trueeyu) [inline be/src/exprs/agg/array_agg.h:226]: = default?

**Review by trueeyu — DISMISSED** [review]

**Review by satanson — APPROVED** [review]

**Review by trueeyu — APPROVED** [review]

**Pull request merged**